lib/lightning_web/plugs/webhook_auth.ex

```diff
@@ -5,76 +5,97 @@ defmodule LightningWeb.Plugs.WebhookAuth do
-  returns a 404 Not Found response with a JSON error message indicating that the webhook
-  is not found.
-
-  ## Parameters
-  - `conn`: The connection struct representing the incoming HTTP request.
-  - `_opts`: A set of options, not used in this function but is a mandatory parameter as per
-     Plug specification.
-
-  ## Returns
-  - A connection struct representing the outgoing response, which can be a successful
-    response, an unauthorized response, or a not found response, based on the evaluation
-    of the above-mentioned conditions.
-
-  ## Examples
-  Assuming a request with the path `/i/some_webhook`:
-
-  ### Webhook Found and Authenticated
-
-      iex> LightningWeb.Plugs.WebhookAuth.call(conn, [])
-      %Plug.Conn{status: 200, ...}
-
-      iex> LightningWeb.Plugs.WebhookAuth.call(conn, [])
-      %Plug.Conn{status: 404, ...}
+  Handles webhook auth for `/i/:webhook` requests.
+
+  - **CORS preflight:** If the request method is `OPTIONS`, this plug is a no-op
+    and returns the connection unchanged so upstream CORS handling can respond.
+    This avoids doing DB lookups or emitting 401/404 on preflight requests.
+
+  - **Auth flow:** For non-`OPTIONS` requests whose path matches `/i/:webhook`,
+    this plug:
+      1. Looks up the webhook trigger (with `workflow` and `edges`) and its
+         `webhook_auth_methods`, wrapped in `Lightning.Retry.with_webhook_retry/2`
+         so transient DB errors are retried.
+      2. If the trigger is missing → responds **404** `{"error":"webhook_not_found"}`.
+      3. If auth methods are configured:
+         - If credentials match → assigns `:trigger` and continues.
+         - If credentials are present but wrong → responds **404** (hide existence).
+         - If credentials are missing → responds **401**.
+      4. If retries exhaust due to DB issues → responds **503** with `Retry-After`
+         based on `WEBHOOK_RETRY_TIMEOUT_MS`.
+
+  Returns the (possibly halted) connection.
   """
+  @spec call(Plug.Conn.t(), any) :: Plug.Conn.t()
+  def call(%Plug.Conn{method: "OPTIONS"} = conn, _opts), do: conn
+
   def call(conn, _opts) do
     case conn.path_info do
-      ["i" | [webhook | _additional_path]] ->
-        trigger =
-          Workflows.get_webhook_trigger(webhook,
-            include: [:workflow, :edges]
-          )
-
-        validate_auth(trigger, conn)
+      ["i" | [webhook | _rest]] ->
+        Retry.with_webhook_retry(
+          fn ->
+            trigger =
+              Workflows.get_webhook_trigger(webhook,
+                include: [:workflow, :edges, :webhook_auth_methods]
+              )
+
+            methods = (trigger && trigger.webhook_auth_methods) || []
+            {:ok, {trigger, methods}}
+          end,
+          retry_on: &Retry.retriable_error?/1,
+          context: %{op: :webhook_auth_lookup, webhook: webhook}
+        )
+        |> case do
+          {:ok, {trigger, methods}} ->
+            validate_auth(trigger, methods, conn)
```

Comment:
Can't this just be part of the `fn` block? I think the success case should be inside the retry block and then  we only handle `error` cases

Reply:
Yes it's actually nicer to have this in the `fn` block, thanks for the suggestion.